service/src/main/java/org/databiosphere/workspacedataservice/expressions/ExpressionService.java

```diff
@@ -0,0 +1,530 @@
+              return Map.entry(
+                  expressionEntry.getKey(), expressionValueSubstitutionVisitor.visit(parsedTree));
+            })
+        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue));
+  }
+
+  /**
+   * Generate a map of lookup text -> expression result for each expression. The expression result
+   * is a JsonNode which can be a null node, a single value, or an array of values.
+   *
+   * <ul>
+   *   <li>null node when query info indicates that the results should not be an array AND
+   *       <ul>
+   *         <li>the results contain an empty list of records OR
+   *         <li>the attribute does not exist on any record
+   *       </ul>
+   *   <li>single value when query info indicates that results should not be an array AND the
+   *       results contain a single record with the attribute
+   *   <li>array of 0 or more elements when query info indicates that the attribute is an array
+   * </ul>
+   *
+   * @param resultsByQuery A map of expression query info -> query result. Each query info may
+   *     contain multiple attribute lookups and each lookup has its own text value which will be a
+   *     key in the result.
+   * @param recordType The record type the expressions were evaluated against
+   */
+  Map<String, JsonNode> getExpressionResultLookupMap(
+      Map<ExpressionQueryInfo, List<Record>> resultsByQuery, RecordType recordType) {
+    return resultsByQuery.entrySet().stream()
+        .flatMap(
+            queryInfoAndResult ->
+                queryInfoAndResult.getKey().attributeLookups().stream()
+                    .map(
+                        lookup -> {
+                          var attributeValues =
+                              queryInfoAndResult.getValue().stream()
+                                  .map(
+                                      record ->
+                                          lookupAttributeValue(
+                                              recordType,
+                                              lookup,
+                                              record,
+                                              queryInfoAndResult.getKey().relations()))
+                                  .filter(Objects::nonNull)
+                                  .map(this::toJsonNode)
+                                  .toList();
+                          if (queryInfoAndResult.getKey().isArray()) {
+                            return Map.entry(
+                                lookup.lookupText(),
+                                objectMapper.createArrayNode().addAll(attributeValues));
+                          } else if (attributeValues.size() > 1) {
+                            throw new ResponseStatusException(
+                                HttpStatus.BAD_REQUEST,
+                                "Expected a single value for attribute %s but got %s"
+                                    .formatted(lookup.attribute(), attributeValues));
```

Comment:
consider returning the count of values found instead of the values themselves:
```suggestion
                                "Expected a single value for attribute %s but got %s"
                                    .formatted(lookup.attribute(), attributeValues.size()));
```